diff --git a/lib/esm/index.js b/lib/esm/index.js
index 37099b20ced7f8c37a1d95319d571c854d607252..1115dec25e7e07b70ac0588f2a795190d4e312d5 100644
--- a/lib/esm/index.js
+++ b/lib/esm/index.js
@@ -26,9 +26,9 @@ export const fromPinyinSyllable = (pinyin) => {
   return formattedResult
 }
 
-export const fromPinyin = (input, everything = false) => {
+export const fromPinyin = (input, everything = false, dictMode = false) => {
   const translate = (pinyin) => {
-    return split(pinyin, everything).map((item) => {
+    return (dictMode ? pinyin.split(" ") : split(pinyin, everything)).map((item) => {
       if (everything) {
         if (typeof item === "string") return item
         else {
diff --git a/lib/types/index.d.ts b/lib/types/index.d.ts
index 970c191a424176f8da41e9d16a02a16581177865..df374f749a2c852e02d5d64d1a91a44db51f5781 100644
--- a/lib/types/index.d.ts
+++ b/lib/types/index.d.ts
@@ -2,7 +2,8 @@ export declare const toneMarks: string[]
 export declare const fromPinyinSyllable: (pinyin: string) => string
 export declare const fromPinyin: (
   input: string | string[],
-  everything?: boolean
+  everything?: boolean,
+  dictMode?: boolean
 ) => string[] | string[][]
 export declare const splitZhuyin: (
   zhuyin: string,
diff --git a/package.json b/package.json
index 776a4efb84d06c78f1d12cd62cbaf41d2409ad22..b695afb57f4f202995d7c911d491f208c614d042 100644
--- a/package.json
+++ b/package.json
@@ -24,6 +24,13 @@
   "main": "./lib/cjs/index.js",
   "module": "./lib/esm/index.js",
   "types": "./lib/types/index.d.ts",
+  "exports": {
+    ".": {
+      "import": "./lib/esm/index.js",
+      "require": "./lib/cjs/index.js",
+      "types": "./lib/types/index.d.ts"
+    }
+  },
   "files": [
     "lib/"
   ],
